### ISSUE TITLE
CloudFormation YAML short form !Sub / Fn::Sub in dependent resource causes dependency resolution failure

### ISSUE THREAD

**Issue by m-chandler** [localstack/localstack#3251]
# Type of request: This is a ...
[ X ] bug report
[ ] feature request

# Detailed description
It appears that if you use the short form CloudFormation Sub in a dependent resource, the CloudFormation stack fails to deploy (e.g. use it in the LogGroupName property of a LogGroup, which a Lambda function then depends on).

I am trying to deploy a Lambda function with a dependency on an LogGroup. I'd also like to parameterise my log group, with an environment name.

If the LogGroup resource's LogGroupName property is a static string, all works well (e.g. dev-log-group).

Workaround: Do not use short form Sub YAML (but I would like to :-) ).

Example template is below. Note the repeated `LogGroupName` properties. The comments indicate which variants work / which don't.

```
AWSTemplateFormatVersion: '2010-09-09'
Description: Template which showcases dependency resolution failure with shorthand Sub
Parameters:
  Environment:
    Type: String
    Description: Environment name
    Default: 'Test'
Resources:
  LogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      # This does not work
      LogGroupName: !Sub "aws/lambda/${Environment}-function"
      # This also does not work
      LogGroupName: 
        Fn::Sub: "aws/lambda/${Environment}-function"
      # This does work
      LogGroupName:
        Fn::Sub: 
          - "aws/lambda/${Environment}-function"
          - {}
      # This works too
      LogGroupName:
        Fn::Join:
        - ''
        - - aws/lambda/
          - !Ref Environment
          - -function
  Role:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
        - Effect: Allow
          Principal:
            Service:
            - lambda.amazonaws.com
          Action:
          - sts:AssumeRole
      Policies:
      - PolicyName: root
        PolicyDocument:
          Version: '2012-10-17'
          Statement:
          - Effect: Allow
            Action:
            - logs:CreateLogStream
            - logs:CreateLogGroup
            - logs:PutLogEvents
            Resource: '*'
      Path: "/"
      RoleName: my-lambda-role
  LambdaFunction:
    Type: AWS::Lambda::Function
    DependsOn: 
    - LogGroup
    - Role
    Properties:
      Code:
        ZipFile: |
          def handler(event, context):
            print("Hello world")
      FunctionName: function
      Handler: index.handler
      MemorySize: 245
      Runtime: python3.8
      Role:
        Fn::GetAtt:
        - Role
        - Arn
      Timeout: 15


```

## Expected behavior

Short form YAML should work.

## Actual behavior

The stack fails to deploy.

Errors in the logs:

```
localstack_main   | 2020-11-17T05:11:49:DEBUG:localstack.utils.cloudformation.template_deployer: Dependency for resource LambdaFunction not yet deployed: LogGroup {'Type': 'AWS::Logs::LogGroup', 'Properties': {'LogGroupName': {'Fn::Sub': 'aws/lambda/${Environment}-function'}}}
localstack_main   | 2020-11-17T05:11:49:INFO:localstack.services.cloudformation.cloudformation_starter: Resource LambdaFunction cannot be deployed, found unsatisfied dependencies. ['LogGroup'] - {'Type': 'AWS::Lambda::Function', 'DependsOn': ['LogGroup', 'Role'], 'Properties': {'Code': {'ZipFile': 'def handler(event, context):\n  print("Hello world")\n'}, 'FunctionName': 'function', 'Handler': 'index.handler', 'MemorySize': 245, 'Runtime': 'python3.8', 'Role': {'Fn::GetAtt': ['Role', 'Arn']}, 'Timeout': 15}} (existing: {'AWS::Region', 'LogGroup', 'AWS::StackId', 'AWS::StackName', 'AWS::Partition', 'AWS::AccountId', 'Role', 'Environment', 'AWS::URLSuffix'})
localstack_main   | 2020-11-17T05:11:49:INFO:localstack.services.cloudformation.cloudformation_starter: Unresolvable dependencies, there may be undeployed stack resources: {'LambdaFunction': ['LogGroup']}
localstack_main   | 2020-11-17T05:11:49:WARNING:bootstrap.py: Thread run method <function apply_patches.<locals>.run_dependencies_deployment_loop.<locals>.run_loop at 0x7fccb17fadc0>(None) failed: Unable to resolve all CloudFormation resources after traversing dependency tree (maximum depth 40): ['LambdaFunction'] Traceback (most recent call last):
localstack_main   |   File "/opt/code/localstack/localstack/utils/bootstrap.py", line 578, in run
localstack_main   |     result = self.func(self.params)
localstack_main   |   File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 1093, in run_loop
localstack_main   |     raise Exception('Unable to resolve all CloudFormation resources after traversing ' +
localstack_main   | Exception: Unable to resolve all CloudFormation resources after traversing dependency tree (maximum depth 40): ['LambdaFunction']
```
    
# Steps to reproduce
    
## Command used to start LocalStack
    
`docker-compose up`

```
version: '2.1'
services:
  localstack:
    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
    image: localstack/localstack
    ports:
      - "4566:4566"
      - "4510:4510"
      - "4571:4571"
      - "4572:4572"
      - "4513:4513"
      - "4514:4514"
      - "${PORT_WEB_UI-8080}:${PORT_WEB_UI-8080}"
    environment:
      - SERVICES=${SERVICES- }
      - DATA_DIR=${DATA_DIR- }
      - PORT_WEB_UI=${PORT_WEB_UI- }
      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR- }
      - KINESIS_ERROR_PROBABILITY=${KINESIS_ERROR_PROBABILITY- }
      - DOCKER_HOST=unix:///var/run/docker.sock
      - HOST_TMP_FOLDER=${TMPDIR}
      - LOCALSTACK_API_KEY=...
      - DEBUG=1
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock"
    tmpfs:
      - /tmp/localstack:exec,mode=600
    network_mode: bridge
```

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)
Save my example template as `dependency-example.yml`
Comment out the LogGroupName properties that aren't under test.
Deploy `awslocal cloudformation create-stack --stack-name dependency-example --template-body file://dependency-example.yml --parameters ParameterKey=Environment,ParameterValue=dev`

┆Issue is synchronized with this [Jira Bug](https://localstack.atlassian.net/browse/LOC-328) by [Unito](https://www.unito.io/learn-more)

**Comment by whummer** [comment]
Thanks for reporting @m-chandler . This should be fixed in #3286 - can you please pull the `latest` Docker image and give it another try? Please report here if the problem persists. Thanks

**Comment by m-chandler** [comment]
Perhaps this issue has been fixed, however I'm still getting a lot of issues with dependency resolution when intrinsic functions are used. It would appear that if Join/Sub is used in a Lambda function name, and that function is then depended on by another resource, the dependency cannot be resolved. I will create another issue which clearly outlines the fault.
